include/swift/AST/Decl.h

```diff
@@ -184,6 +184,25 @@ enum class StaticSpellingKind : uint8_t {
   KeywordClass,
 };
 
+/// Describes the kind of implicit constructor that will be
+/// generated.
+enum class ImplicitConstructorKind : uint8_t {
+  /// The default constructor, which default-initializes each
+  /// of the instance variables.
+  Default,
+  /// The memberwise constructor, which initializes each of
+  /// the instance variables from a parameter of the same type and
+  /// name.
+  Memberwise,
+  /// An implicit constructor imported from Objective-C.
+  Imported,
```

Comment:
I don't think we really need to keep track of the Imported, SynthesizedProtocolRequirement or Designated cases.